### PR TITLE
HPCC-19431 Use Bloom filter to avoid btree lookups

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Regression suite and unit tests. Also verified using debugger that the code to skip a lookup did actually happen in some cases.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Perhaps worth an initial look, though some cleaning up needed

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 1]
```diff
@@ -136,10 +136,8 @@ class FilterSet
         if (fieldNum == (unsigned) -1)
             throw MakeStringException(0, "Invalid filter string: field '%s' not recognized", fieldName.str());
         unsigned numOffsets = inrec.getNumVarFields() + 1;
-        size_t * variableOffsets = (size_t *)alloca(numOffsets * sizeof(size_t));
-        RtlRow offsetCalculator(inrec, nullptr, numOffsets, variableOffsets);
-        unsigned fieldOffset = offsetCalculator.getOffset(fieldNum);
-        unsigned fieldSize = offsetCalculator.getSize(fieldNum);
+        unsigned fieldOffset = inrec.getFixedOffset(fieldNum);
```
Comment (richardkchapman) [inline rtl/eclrtl/eclhelper_dyn.cpp]: This change not directly related, but found in testing that the filter feature in dumpkey had been broken by the changes to lazy field offset calculation. Code was a bit dodgy before and is equally dodgy now.

File: system/jlib/jarray.hpp [hunk 2]
```diff
@@ -27,7 +27,7 @@
 
 typedef size32_t aindex_t;
 
-#define NotFound   (aindex_t)-1
```
Comment (richardkchapman) [inline system/jlib/jarray.hpp]: Was clashing with something in 3rd-patry cuckoo filter code I was trying out. Decided not to use it but this change is still a good one I think.

File: system/jhtree/keybuild.cpp [hunk 3]
```diff
@@ -490,11 +515,40 @@ class CKeyBuilder : public CKeyBuilderBase, implements IKeyBuilder
...
+            nextPos += keyHdr->getNodeSize();
+            size32_t written = node->set(data, size);
+            assertex(written);
+            if(prevNode)
+            {
+                node->setLeftSib(prevNode->getFpos());
+                prevNode->setRightSib(node->getFpos());
+                writeNode(prevNode);
+            }
+            prevNode.setown(node.getLink());
```
Comment (ghalliday) [inline system/jhtree/keybuild.cpp]: prevNode.setown(node.getClear()) is better.

File: system/jhtree/CMakeLists.txt [hunk 4]
```diff
@@ -40,12 +41,14 @@ set (    INCLUDES
...
          jhtree.hpp
          jhutil.hpp
+         bloom.hpp
          keybuild.hpp
          keydiff.hpp
     )
 
 include_directories ( 
          .
+         ./cuckoofilter/src
```
Comment (ghalliday) [inline system/jhtree/CMakeLists.txt]: no longer needed

File: system/jhtree/bloom.cpp [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+#include "math.h"
+
+BloomFilter::BloomFilter(unsigned _cardinality, double _probability)
+{
+    unsigned cardinality = _cardinality ? _cardinality : 1;
+    double probability = _probability >= 0.3 ? 0.3 : (_probability < 0.01 ? 0.01 : _probability);
+    numBits = rtlRoundUp(-(cardinality*log(probability))/pow(log(2),2));
+    unsigned tableSize = (numBits + 7) / 8;
+    numBits = tableSize * 8;
+    numHashes = (numBits/cardinality)*log(2);
```
Comment (ghalliday) [inline system/jhtree/bloom.cpp]: numBits/cardinality should be done as a real operation.

File: system/jhtree/jhtree.cpp [hunk 6]
```diff
@@ -1539,6 +1573,36 @@ offset_t CKeyIndex::queryMetadataHead()
...
+    if (!bloomAddr || bloomAddr == static_cast<offset_t>(-1))
+        return; // indexes created before introduction of bloomfilter would have FFFF... in this space
+    uint32_t bloomTableSize = keyHdr->getHdrStruct()->bloomTableSize;
+    unsigned short bloomTableHashes = keyHdr->getHdrStruct()->bloomTableHashes;
+    Owned<CJHTreeBloomTableNode> node;
+    MemoryBuffer bloomTable;
+    bloomTable.ensureCapacity(bloomTableSize);
+    while (bloomAddr)
+    {
+        node.setown(QUERYINTERFACE(loadNode(bloomAddr), CJHTreeBloomTableNode));
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: static cast would probably be better, possibly following an assert that the type matches what you expect.

File: system/jhtree/keybuild.cpp [hunk 7]
```diff
@@ -336,20 +337,28 @@ class CKeyBuilder : public CKeyBuilderBase, implements IKeyBuilder
...
-        : CKeyBuilderBase(_out, flags, rawSize, nodeSize, keyedSize, startSequence), enforceOrder(_enforceOrder)
+    CKeyBuilder(IFileIOStream *_out, unsigned flags, unsigned rawSize, unsigned nodeSize, unsigned keyedSize, unsigned __int64 startSequence, unsigned _bloomKeyLength, bool _enforceOrder)
+        : CKeyBuilderBase(_out, flags, rawSize, nodeSize, keyedSize, startSequence), bloomKeyLength(_bloomKeyLength), enforceOrder(_enforceOrder)
     {
         doCrc = true;
         activeNode = NULL;
         activeBlobNode = NULL;
         duplicateCount = 0;
+        if (bloomKeyLength)
+            lastKeyData = (byte *) calloc(bloomKeyLength, 1);
```
Comment (ghalliday) [inline system/jhtree/keybuild.cpp]: lastBloomData would be a less confusing name.

File: system/jhtree/keybuild.cpp [hunk 8]
```diff
@@ -413,6 +427,17 @@ class CKeyBuilder : public CKeyBuilderBase, implements IKeyBuilder
             if (cmp==0)
                 ++duplicateCount;
         }
+        if (bloomKeyLength)
+        {
+            int cmp = memcmp(keyData, lastKeyData, bloomKeyLength);
+            if (cmp)
+            {
+                memcpy(lastKeyData, keyData, bloomKeyLength);
+                hash64_t hash = rtlHash64Data(bloomKeyLength, keyData, 0);
```
Comment (ghalliday) [inline system/jhtree/keybuild.cpp]: Should probably use a different seed from 0 e.g. HASH64_INIT

File: system/jhtree/bloom.hpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+#include "jhtree.hpp"
+
+class jhtree_decl BloomFilter : public CInterface
+{
+public:
+    BloomFilter(unsigned cardinality, double probability=0.1);
+    BloomFilter(unsigned numHashes, unsigned tableSize, byte *table);
+    void add(hash64_t hash);
+    bool test(hash64_t hash) const;
+    inline void add(size32_t len, const void *val) { add(rtlHash64Data(len, val, 0)); }
```
Comment (ghalliday) [inline system/jhtree/bloom.hpp]: Better to use HASH64_INIT as the seed.

File: system/jhtree/bloom.hpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+{
+public:
+    BloomBuilder();
+    const BloomFilter * build(double probability=0.1) const;
+    bool add(hash64_t hash);
+    inline bool add(size32_t len, const void *val) { return add(rtlHash64Data(len, val, 0)); }
+    bool valid() const;
+protected:
+    ArrayOf<hash64_t> hashes;
+    const unsigned maxHashes = 1000000;  // MORE - allow user to override?
```
Comment (ghalliday) [inline system/jhtree/bloom.hpp]: Should be an option on the constructor

**Comment by richardkchapman** [comment]
@ghalliday Please re-review

File: system/jhtree/bloom.hpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+#define _BLOOM_INCL
+
+#include "jhtree.hpp"
+
+class jhtree_decl BloomFilter : public CInterface
+{
+public:
+    BloomFilter(unsigned cardinality, double probability=0.1);
+    BloomFilter(unsigned numHashes, unsigned tableSize, byte *table);
+    void add(hash64_t hash);
```
Comment (ghalliday) [inline system/jhtree/bloom.hpp]: BloomFilter leaks table on destruction..

**Comment by richardkchapman** [comment]
@ghalliday Repushed to add comments and fix leak

**Pull request merged**